src/main/io/gps.c

```diff
@@ -843,6 +1368,7 @@ void gpsUpdate(timeUs_t currentTimeUs)
 
     // read out available GPS bytes
     if (gpsPort) {
+        DEBUG_SET(DEBUG_GPS_CONNECTION, 7, serialRxBytesWaiting(gpsPort));
         while (serialRxBytesWaiting(gpsPort)) {
             if (cmpTimeUs(micros(), currentTimeUs) > GPS_MAX_WAIT_DATA_RX) {
```

Comment:
calling `micros()` for every byte read is rather expensive.
Replacing with the code below to check time after 10 bytes read yields a 300x performance increase (as simulated on a desktop)
```
int check_every = 10;
if (check_every-- <= 0 && cmpTimeUs(micros(), currentTimeUs) > GPS_MAX_WAIT_DATA_RX) {
```

Reply:
![Screen Shot 2023-07-29 at 23 26 23](https://github.com/betaflight/betaflight/assets/11737748/bba36cb0-cd64-4807-8a18-35287fb7ac5a)
This graphic can be compared to others previously recorded.
Debug 7 (pink) is the buffer size; it peaks around 233 with the Sat package arriving every 5th Nav package while in Configurator (M10, 115200, buffer 256).
The peaks are about the same magnitude as before, but I think the buffer clears a bit more quickly with this change.
![Screen Shot 2023-07-29 at 23 33 12](https://github.com/betaflight/betaflight/assets/11737748/4d467b8f-f855-4918-9145-490b94280bd1)

Reply:
That does look clean, without the timeout check.
![Screen Shot 2023-07-30 at 00 53 57](https://github.com/betaflight/betaflight/assets/11737748/e3c9a9c0-cfc8-40c0-9687-a31b0ecceeed)

Reply:
38400 is about as low as you can go at 10hz, either way
![Screen Shot 2023-07-30 at 01 35 22](https://github.com/betaflight/betaflight/assets/11737748/7dc9f54b-8ca4-49df-b1d5-5592720d18f4)

Reply:
These are really good screenshots. I would advocate for a bit more pessimism and turning on (manually) more messages than we need to compare the two approaches.
Sure, in ideal circumstances these above represent what we process, but in some other circumstances (faster GPS rate, lost of unneeded messages, etc) the GPS module might hammer us with an actual 115200bps data (the current amount is very little compared to what can come at max speed).

Reply:
I think its OK.  I've been testing with the overhead of a USB connection and all sat messages enabled.
In flight there won't be any Sat messages, normally, but even if they existed, it worked fine with this level of testing.

If people use modules that send a ton of junk, and won't stop, we'll hear about those situations.
For now I think the testing is sufficient to be reasonably confident that there are no issues with overflow at this point.   I have already tested with very high sat counts, and that wasn't a problem at 115200 or 57600.

I'm thinking that if we use your suggestion to collect chunks of characters, a value smaller than 50 may be better suited for the lower baud rates.  Perhaps 20 or something like that.

Reply:
Just to clarify: you are saying that using the suggestion of `check_every` did not make a difference / did not clear the buffer faster, therefore, no need to implement it?
Or are you saying why implement it if for the time being we did not find a concrete bug on your setup, even when calling `micros()` for every byte we read?
If it's the first, then obviously no reason to implement it. If it's the second, then it's a no-brainer to implement it (less CPU overhead, faster GPS processing)

Reply:
This seems to be only safeguard to prevent processing GPS data for too long (and blocking other tasks). IMO it can be removed for now and some better implementation used eventually. Or simply limit number of bytes processed on each iteration

Reply:
> calling `micros()` for every byte read is rather expensive. Replacing with the code below to check time after 50 bytes read yields a substantial performance increase (as simulated on a desktop)
> 
> ```
> int check_every = 50;
> while (serialRxBytesWaiting(gpsPort)) {
>  if (check_every-- <= 0 && cmpTimeUs(micros(), currentTimeUs) > GPS_MAX_WAIT_DATA_RX) {
> ```
> 
> This would likely solve all buffer overrun problems as well as the parsing code would have much more time to process the incoming bytes (rather than doing arithmetic so calculate `micros()`)
> 
> EDIT: it does not seem to matter much if we check every 100 or 50 bytes, so it's probably safer to keep it lower for very slow FCs
> 
> EDIT2: I would actually advocate for removing the `micros()` check altogether:
> 
> * The rest of the GPS code is not time-checked anyway, not much point in perfecting the timing of byte-reading but then ignoring how long the parsing takes
> * We know what speed we need to support: 115200bps=12800byte/s=12.8byte/schedule at 1KHz, as we reschedule the buffer reading if we are not done 1ms later.
> 
> Therefore, bailing and rescheduling after 13-20 chars (rather than worrying about `micros`) is the optimal approach if we don't care about initial fill-up: the usual rate is every 10ms (100Hz), in which we can receive 128 bytes even before we start reading anything. `check_every = 50` it is without an extra `micros()` check

The 30us check is to ensure the gosUpdate() task plays will in the cooperative scheduling scheme we use. If there’s more work to be done we break it down into 30us chunks. It’s there for good reason. Running to completion will break system determinism.

Reply:
It's with good intentions, but with a poor implementation, @SteveCEvans as the CPU will spend *most* of its time doing the arithmetic needed for `micros()`, calling it for every single byte read, rather than actual work. A better implementation is needed here.

Reply:
Could we maybe set `check_every` to 25, and then once 25 values are accepted, do the millis() check?

Reply:
This should be resolved by [commit](https://github.com/betaflight/betaflight/pull/12799/commits/2edf61d6cc2d6815f8dfcc469ec69c5b6cc5e713)